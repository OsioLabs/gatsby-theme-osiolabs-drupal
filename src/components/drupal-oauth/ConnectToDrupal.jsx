import React from 'react';
import PropTypes from 'prop-types';
import { OauthSender } from 'react-oauth-flow';

const DrupalConnectLink = ({ url, queryString, title, classes }) => (
  <a href={`${url}${queryString}`} className={classes}>
    {title}
  </a>
);

/**
 * Create a link that can be used for registration.
 *
 * This takes the authorize URL generated by OauthSender, extracts the query
 * string from it, and then uses that to generate a new URL for /user/register
 * with an ?destination parameter that will redirect to the authorize URL with
 * the appropriate query string after the registration process is complete.
 *
 * This takes advantage of some Drupalisms like ?destination=.
 */
const DrupalRegisterLink = ({ url, queryString, title, drupalRoot, classes }) => {
  let query = url.split('?')[1];
  query = encodeURIComponent(`?${query}`);
  return (
    <a
      href={`${drupalRoot}/user/register?destination=/oauth/authorize${query}${queryString}`}
      className={classes}
    >
      {title}
    </a>
  );
};

const ConnectToDrupal = props => {
  // Normalize the redirect URI so that it always contains a trailing slash.
  // Gatsby likes to redirect example.com/asdf?code= to example.com/asdf/?code=
  // and if the URL that we send in the original request doesn't match the one
  // we use to verify the ?code= param there will be problems.
  let redirectUri = props.redirectUri;
  redirectUri += redirectUri.endsWith('/') ? '' : '/';

  return (
    <OauthSender
      authorizeUrl={props.drupalOauthClient.config.authorize_url}
      clientId={props.drupalOauthClient.config.client_id}
      redirectUri={redirectUri}
      state={{}}
      render={({ url }) => {
        if (props.useRegistrationLink === true) {
          return (
            <DrupalRegisterLink
              url={url}
              queryString={props.queryString}
              title={props.title}
              classes={props.classes}
              drupalRoot={props.drupalOauthClient.config.drupal_root}
            />
          );
        }
        return (
          <DrupalConnectLink
            url={url}
            queryString={props.queryString}
            title={props.title}
            classes={props.classes}
          />
        );
      }}
    />
  );
};

ConnectToDrupal.propTypes = {
  // An instance of the DrupalOauth class, used to obtain configuration.
  drupalOauthClient: PropTypes.object.isRequired,
  // The URL to redirect to after obtaining a grant from the server.
  redirectUri: PropTypes.string.isRequired,
  // The text to use for the link.
  title: PropTypes.string.isRequired,
  // If set to true use a link that will allow someone to register a new
  // account, and then authorize the grant.
  useRegistrationLink: PropTypes.bool,
  // Additional query string parameters to add to the link.
  queryString: PropTypes.string,
  // A list of class names to apply to the a element.
  classes: PropTypes.string,
};

ConnectToDrupal.defaultProps = {
  useRegistrationLink: false,
  queryString: '',
  classes: '',
};

export default ConnectToDrupal;
