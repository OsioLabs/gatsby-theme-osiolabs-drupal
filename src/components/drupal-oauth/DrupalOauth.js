import retry from 'async-retry';
import qs from 'qs';

class DrupalOauth {
  processing = false;

  /**
   *
   * @param config
   * - drupal_root:
   * - token_url:
   * - authorize_url:
   * - client_id:
   * - client_secret:
   */
  constructor(config) {
    this.config = config;
    this.config.token_url = `${this.config.drupal_root}/oauth/token`;
    this.config.authorize_url = `${this.config.drupal_root}/oauth/authorize`;
  }

  /**
   * Check to see if the current user is logged in.
   *
   * If the user was previously logged in but their access token is expired
   * attempt to retrieve a new token.
   *
   * @param {function} refreshComplete
   *  Optional callback function to call after a refresh token has been
   *  processed.
   *
   * @returns <Mixed>
   *   The current users authorization token, or false.
   */
  isLoggedIn = async refreshComplete =>
    retry(async bail => {
      if (
        typeof localStorage === 'undefined' ||
        typeof window === 'undefined'
      ) {
        // This is being executed outside of the browser, so it's not going to work
        // and we can just bail out now
        bail();
        return false;
      }

      const token =
        localStorage.getItem('drupal-oauth-token') !== null
          ? JSON.parse(localStorage.getItem('drupal-oauth-token'))
          : null;

      // If there is no token, and we're not processing, that means there is no
      // logged in user. So we're done.
      if (token === null && this.processing === false) {
        bail();
        return false;
      }

      // If we're currently trying to retrieve a refersh token from another
      // request then just retry.
      if (this.processing === true) {
        throw new Error('retry');
      }

      // If we've got an active token, the user is logged in, so we return their
      // current token.
      if (
        token !== null &&
        token.expirationDate > Math.floor(Date.now() / 1000)
      ) {
        this.processing = false;
        return token;
      }

      // If there is an existing token, but it's expired, see if we can get a
      // a new one using the refresh token. If not, see if we can get a refresh
      // token.
      this.processing = true;
      await this.getRefreshToken(token).then(fullfilledToken => {
        if (fullfilledToken !== null) {
          this.processing = false;
          if (typeof refreshComplete === 'function') {
            refreshComplete(true);
          }
          return fullfilledToken;
        }

        bail(new Error('Unable to retrieve an access token'));
        return false;
      });

      return false;
    });

  /**
   *
   */
  async handleLogin(username, password, scope) {
    return this.fetchOauthToken(username, password, scope);
  }

  /**
   * Log the current user out.
   *
   * Deletes the token from local storage.
   */
  // eslint-disable-next-line class-methods-use-this
  async handleLogout() {
    return localStorage.removeItem('drupal-oauth-token');
  }

  async getOauthToken(username, password, scope) {
    return this.fetchOauthToken(username, password, scope);
  }

  async getRefreshToken(token) {
    return this.refreshOauthToken(token);
  }

  /**
   * Get an OAuth token from Drupal.
   *
   * Exchange a an authorization code for an OAuth token. Uses ?code and other
   * parameters from the current URL.
   *
   * @returns {Promise<void>}
   *   Returns a promise that resolves with the new token returned from Drupal.
   */
  // eslint-disable-next-line consistent-return
  async fetchOauthToken(url) {
    this.processing = true;

    const queryString = qs.parse(url.split('?')[1], {
      ignoreQueryPrefix: true,
    });

    if (!queryString.code) {
      throw new Error('Can not retrieve authorization token without a code.');
    }

    const formData = new FormData();
    Object.keys(queryString).forEach(item => {
      formData.append(item, queryString[item]);
    });

    const response = await fetch(this.config.token_url, {
      method: 'post',
      headers: new Headers({
        Accept: 'application/json',
      }),
      body: formData,
    });

    if (response.ok) {
      const json = await response.json();

      if (json.error) {
        this.processing = false;
        throw new Error(json.error.message);
      }

      this.processing = false;
      return this.storeToken(json);
    }
  }

  /**
   * Exchange your refresh token for a new auth token.
   *
   * @param token
   * @param scope
   *
   * @returns {Promise<void>}
   *  Returns a Promise that resolves with the new token retrieved from Drupal.
   */
  // eslint-disable-next-line consistent-return
  async refreshOauthToken(token) {
    if (token !== null) {
      const formData = new FormData();
      formData.append('grant_type', 'refresh_token');
      formData.append('client_id', this.config.client_id);
      formData.append('client_secret', this.config.client_secret);
      formData.append('refresh_token', token.refresh_token);

      const response = await fetch(this.config.token_url, {
        method: 'post',
        headers: new Headers({
          Accept: 'application/json',
          'X-Consumer-ID': `${this.config.client_id}`,
        }),
        body: formData,
      });

      if (response.ok) {
        const json = await response.json();

        if (json.error) {
          throw new Error(json.error.message);
        }

        return this.storeToken(json);
      }

      throw new Error(response.status);
    }
  }

  // eslint-disable-next-line class-methods-use-this
  storeToken(json) {
    const token = Object.assign({}, json);
    token.date = Math.floor(Date.now() / 1000);
    token.expirationDate = token.date + token.expires_in;
    localStorage.setItem('drupal-oauth-token', JSON.stringify(token));
    return token;
  }
}

export default DrupalOauth;
