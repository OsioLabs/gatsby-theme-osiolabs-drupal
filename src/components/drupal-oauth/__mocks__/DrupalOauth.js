// __mocks__/DrupalOauth.js

const mock = jest.fn().mockImplementation(() => ({
  config: {
    drupal_root: 'http://api.example.com',
    authorize_url: 'http://api.example.com/oauth/authorize',
    token_url: 'http://api.example.com/oauth/token',
    client_id: 'CLIENTID',
    client_secret: 'CLIENTSECRET',
  },
  isLoggedIn: jest.fn(() =>
    Promise.resolve({ access_token: 'access-token', token_type: 'Bearer' })
  ),
  handleLogin: jest.fn(() => Promise.resolve({})),
  handleLogout: jest.fn(() => Promise.resolve({})),
  fetchOauthToken: jest.fn(() => Promise.resolve({})),
  storeToken: jest.fn(),
}));

mock.createFromEnvironment = jest.fn().mockImplementation(mock);

export default mock;
