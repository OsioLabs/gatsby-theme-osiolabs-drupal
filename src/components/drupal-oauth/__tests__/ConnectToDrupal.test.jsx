/* eslint-disable react/prop-types */
import React from 'react';
import { render, cleanup } from '@testing-library/react';
import DrupalOauth from '../DrupalOauth';
import ConnectToDrupal from '../ConnectToDrupal';

jest.mock('../DrupalOauth');

const drupalOauthClient = new DrupalOauth();

describe('Component: <SendToDrupal />', () => {
  afterEach(cleanup);

  it('Renders a link to authorize a user against Drupal', () => {
    const { getByText, rerender } = render(
      <ConnectToDrupal
        drupalOauthClient={drupalOauthClient}
        title="Connect to Drupal"
        redirectUri="http://www.example.com"
        queryString="&foo=bar"
      />
    );

    const render1 = getByText('Connect to Drupal');
    expect(render1).toMatchSnapshot();

    // Render a second time, NOTE the trailing slash on the redirectUri.
    // This verifies our logic to handle normalizing the redirectUri is
    // working.
    rerender(
      <ConnectToDrupal
        drupalOauthClient={drupalOauthClient}
        title="Connect to Drupal"
        redirectUri="http://www.example.com/"
      />
    );
    const render2 = getByText('Connect to Drupal');
    expect(render1).toEqual(render2);
  });

  it('Renders a link to register and then authorize a user against Drupal', () => {
    const { getByText } = render(
      <ConnectToDrupal
        drupalOauthClient={drupalOauthClient}
        title="Register with Drupal"
        redirectUri="http://www.example.com"
        useRegistrationLink
      />
    );
    expect(getByText('Register with Drupal')).toMatchSnapshot();
  });
});
