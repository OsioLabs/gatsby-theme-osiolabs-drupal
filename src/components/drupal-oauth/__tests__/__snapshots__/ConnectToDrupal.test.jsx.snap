// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <SendToDrupal /> Renders a link to authorize a user against Drupal 1`] = `
<a
  class=""
  href="http://api.example.com/oauth/authorize?client_id=CLIENTID&redirect_uri=http%3A%2F%2Fwww.example.com%2F&response_type=code&state=%7B%7D&foo=bar"
>
  Connect to Drupal
</a>
`;

exports[`Component: <SendToDrupal /> Renders a link to register and then authorize a user against Drupal 1`] = `
<a
  class=""
  href="http://api.example.com/user/register?destination=/oauth/authorize%3Fclient_id%3DCLIENTID%26redirect_uri%3Dhttp%253A%252F%252Fwww.example.com%252F%26response_type%3Dcode%26state%3D%257B%257D"
>
  Register with Drupal
</a>
`;
